### DISCUSSION TITLE
[mini.pick] git worktree picker

### DISCUSSION THREAD

**Discussion by felpafel** [nvim-mini/mini.nvim#1131] [Show and tell]
https://github.com/user-attachments/assets/b16c69c6-88b6-4105-a30e-9d67adfdd5ad

Hello, I've created this picker around 3 months ago, and decided to share it here.

This picker list all available worktrees and change neovim working directory (`cwd`) to the selected tree. [screenshot](https://github.com/user-attachments/assets/8f6af9e1-9b69-40f9-94d2-20607f9c1cde)

I choose to only change the directory so it's simple to anyone to implement their own automation. In my case, I use `DirChangedPre` event and `mini.sessions` to save my current session and load the session to the new working dir (using `oil.nvim` as a fallback in case I don't have a previous session).

### Usage:

The picker accepts an optional `cwd` parameter in case you want to search a repository in another path.

- On the cmdline: `:Pick git_worktrees` or `Pick git_worktrees cwd='my/git/repo/path'`;
- On keymaps: `vim.keymap.set('n', '<leader>fw', '<cmd>Pick git_worktrees<cr>')`

### Implementation:

Basically, this picker queries all available [`git worktrees`](https://git-scm.com/docs/git-worktree) using `git -C $cwd worktree list` and then the cli stdout is parsed using `string.match` to build a table array that has three fields: `text`, `branch` and `path` (you can customize the prompt by changing the content of `text`). So, when the user select an item the `wt_switch` procedure is called, changing the working dir to `item.path`.

### Code:

```lua
local wt_switch = function(item)
  -- NOTE: Change CWD to selected item's path
  vim.schedule(function()
    vim.fn.chdir(item.path)
  end)
  vim.notify(
    'Git worktree changed to ' .. (item.branch and item.branch or '(bare)'),
    vim.log.levels.INFO
  )
end

--- @return { text: string, branch: string, path: string }[]
local wt_postprocess = function(item)
  local items = {}
  for _, entry in pairs(vim.split(item[1], '%z%z', { trimempty = true })) do
    local path, head, _, ref =
    entry:match('^worktree (.-)%zHEAD (.-)%z([^ ]+) (.+)$')
    if path then
      local branch = vim.fs.basename(ref)
      items[#items + 1] = {
        text = string.format(
          '[%s] %s %s',
          branch,
          head:sub(1, 7),
          path
        ),
        branch = branch,
        path = path,
      }
    else
      path = entry:match('^worktree (.-)%z(bare)$')
      if path then
        items[#items + 1] = {
          text = string.format('(bare) %s', path),
          branch = nil,
          path = path,
        }
      end
    end
  end
  return items
end

MiniPick.registry.git_worktrees = function(local_opts)
  local cwd = local_opts.cwd or vim.fn.getcwd()
  if vim.system({ 'git', '-C', cwd, 'rev-parse' }):wait().code ~= 0 then
    vim.notify(
      'Not in a git repository',
      vim.log.levels.ERROR
    )
    return
  end
  local opts = {
    source = {
      -- NOTE: Window Title
      name = string.format('Git Worktrees - %s', cwd),
      choose = wt_switch,
    },
  }
  local_opts.command =
  { 'git', '-C', cwd, 'worktree', 'list', '--porcelain', '-z' }
  local_opts.postprocess = wt_postprocess

  return MiniPick.builtin.cli(local_opts, opts)
end
```

Thanks.

**Comment by echasnovski** [comment]
Thanks for sharing!

I've wanted to try to learn more about Git worktrees for quite a while in order to incorporate them into my workflow. Yet, never really got to it :(

I like the usage of `-z` and zero byte character! It took me a while to understand that this is possible and even preferable to any other ad-hoc delimiter.

**Reply by felpafel** [reply]
I'm glad you like it!

>I've wanted to try to learn more about Git worktrees for quite a while in order to incorporate them into my workflow. Yet, never really got to it :(

Well, I found hard to understand it at first, mostly, because everyone on the internet is trying to sell it as "the killer feature!", instead of "branches as folders".

Using `mini.nvim` as an example, my workflow is the following:

1. `mkdir mini.nvim && cd mini.nvim`
1. `git clone --bare https://github.com/echasnovski/mini.nvim.git `
1. `cd mini.nvim.git`
1. `git worktree add ../main main`
1. `git worktree add ../stable stable`

The final file-tree will be:

```console
 mini.nvim/
 ├─ main/
 ├─ mini.nvim.git/
 └─ stable/
```

In the end, instead of jumping from branch to branch using `git switch` you just change your working directory, and since they are separate folders you don't need to commit, stash or whatever in order to switch.

To wrap up , you only need a worktree if you need to change branches and since git only works inside a git repo, you **can't** use `git` inside the root folder (`mini.nvim`)

https://github.com/user-attachments/assets/a4a3eb8c-ddcb-4047-b3f0-54e64819f861

**Reply by echasnovski** [reply]
Thanks for the explanation and the video (I *do* plan to add 'mini.quickfix', just not right away :) ).

I did think about worktrees as "branches as directories". It's just that "regular" workflow with "stash - checkout" is super straightforward with LazyGit, so I am pretty happy with it.
